sklearn/metrics/tests/test_pairwise.py

```diff
@@ -606,8 +607,273 @@ def test_cosine_distances():
+                                abs(X[i][1] - X[j][1])) / n_cols
+
+    assert_array_almost_equal(D_expected, D)
+
+    # Test X and Y with diferent ranges of numeric values
+    X = [[222.22, 1],
+         [1934.0, 4],
+         [1, 1]]
+
+    Y = [[222.22, 1],
+         [1934.0, 4],
+         [3000, 3000]]
+
+    D = gower_distances(X, Y)
+
+    # The expected normalized values above are:
+    X = [[0.073765, 0.0],
+         [0.644548, 0.001],
+         [0.0,      0.0]]
+
+    Y = [[0.073765, 0.0],
+         [0.644548, 0.001],
+         [1.0,      1.0]]
+
+    # Simplified calculation of Gower distance for expected values
+    n_rows, n_cols = np.shape(X)
+    D_expected = np.zeros((n_rows, n_rows))
+    for i in range(0, n_rows):
+        for j in range(0, n_rows):
+            D_expected[i][j] = (abs(X[i][0] - Y[j][0]) +
+                                abs(X[i][1] - Y[j][1])) / n_cols
+
+    assert_array_almost_equal(D_expected, D)
+
+    # Test to obtain a non-squared distance matrix
+    X = np.array([['Syria', 1.0, 0.0, 0.0, True],
+                  ['Ireland', 0.181818, 0.0, 1, False],
+                  ['United Kingdom', 0.0, 0.0, 0.160377, False]],
+                 dtype=object)
+
+    Y = np.array([['United Kingdom', 0.090909, 0.0, 0.500109, True]],
+                 dtype=object)
+
+    D = gower_distances(X, Y)
+
+    # Simplified calculation of Gower distance for expected values
+    n_rows, n_cols = X.shape[0], Y.shape[0]
+    D_expected = np.zeros((n_rows, n_cols))
+    for i in range(0, n_rows):
+        for j in range(0, n_cols):
+            D_expected[i][j] = \
+                ([1, 0][X[i][0] == Y[j][0]] +
+                 abs(X[i][1] - Y[j][1]) +
+                 abs(X[i][2] - Y[j][2]) +
+                 abs(X[i][3] - Y[j][3]) +
+                 [1, 0][X[i][4] == Y[j][4]]) / X.shape[1]
+
+    assert_array_almost_equal(D_expected, D)
+
+    # Tests a range of negative and positive numeric values
+    # Range starting with zero
+    X = np.array([[0.0], [0.75], [1.0]])
+
+    D = gower_distances(X)
+
+    # Simplified calculation of Gower distance for expected values
+    n_rows, n_cols = np.shape(X)
+    D_expected = np.zeros((n_rows, n_rows))
+    for i in range(0, n_rows):
+        for j in range(0, n_rows):
+            D_expected[i][j] = (abs(X[i][0] - X[j][0])) / n_cols
+
+    assert_array_almost_equal(D_expected, D)
+
+    # Range of positive and negative values
+    X = X - 0.5
+    D = gower_distances(X)
+    assert_array_almost_equal(D_expected, D)
+
+    # Range with positive values
+    X = X + 10
+    D = gower_distances(X)
+    assert_array_almost_equal(D_expected, D)
+
+    # Range of negative values
+    X = X - 15
+    D = gower_distances(X)
+    assert_array_almost_equal(D_expected, D)
+
+    # Test warnings for unexpected non-normalized data
+    X = [[1, 20], [0, -10.0]]
+    with pytest.warns(UserWarning) as record:
+        gower_distances(X, apply_scaling=False)
+        assert len(record) > 0
```

Comment:
we could also test if gower distance returns the Manhattan/Hamming distance for numerical/categorical only input data respectively, I guess.